### DISCUSSION TITLE
Automated Backup to S3

### DISCUSSION THREAD

**Discussion by patrickwilliamconway** [ReagentX/imessage-exporter#265] [Ideas]
First off, I'd like to say thank you for building this tool. This is such a frustrating problem, and I appreciate all the work you've done. The tool has been great to play around with.

I want an archive of all my iMessage data because I don't want to continuously bump my iCloud storage tier. I'd like to use this tool in an automated fashion. I'm imagining running this as a cron that runs every month [week? whatever interval...] and then pushes all output to an S3 bucket that I'll use as long term cold storage since that is significantly cheaper ($0.023/GB/month or [cheaper](https://aws.amazon.com/s3/pricing/)).

Have you thought about supporting an scheduled exporting element to this? I was thinking that this could use something like [AirByte](https://github.com/airbytehq/airbyte) to handle the export.

**Comment by ReagentX** [comment]
You can write your own `cron` job to generate exports and use something like `rsync` to send the data to a cloud service. This feature is out-of-scope.

**Reply by patrickwilliamconway** [reply]
I have a very basic script here (https://github.com/patrickwilliamconway/imessage-backup) that I plan on putting in a crontab and running daily. 

Essentially all I'm doing is:
1. running the tool `imessage-exporter -f txt -o ~/imessage-backup/backups`
2. sync the dir to S3 `aws s3 sync ~/imessage-backup $S3_BUCKET` 

However, I see that the tool doesn't overwriting the target dir:
```shell
Invalid options!
Specified export path "/Users/patrickconway/imessage-backup/backups/" contains existing "txt" export data!
```

To get around this, I just nuke the backups dir, and regenerate it all:
```zsh
  log "Removing all files in $export_dir so they can be regenerated"
  rm -rf ~/imessage-backup/backups
```

and then when I sync files, I'm re-uploading the entire backup again, rather than the diff. The vast majority of these files shouldn't have changed in content.

This isn't _really_ an issue because this runs fast enough and S3 isn't charging for all the PUT and DELETE (afaik, lol). However, would it be possible to support overwriting @ReagentX? I haven't taken a deep look at the source code, so apologies if this is an obvious question.

**Reply by ReagentX** [reply]
This behavior is by design to ensure integrity of the exported data–i.e. accidentally overwriting already-exported files. I would rather users create a new backup directory, or manually remove the existing data, than accidentally do something destructive by mistake.
